torchdata/dataloader2/communication/eventloop.py

```diff
@@ -26,16 +29,62 @@
 
 __all__ = [
     "DataPipeToQueuesLoop",
-    "SpawnProcessForDataPipeline",
-    "SpawnThreadForDataPipeline",
+    "CreateProcessForDataPipeline",
+    "CreateProcessForMultipleDataPipelines",
+    "CreateThreadForDataPipeline",
 ]
 
 
+TIME_SLEEP_BETWEEN_CHECKING_DIFFERENT_QUEUES = 0.00000001
+
+
+def MultipleDataPipesToQueuesLoop(source_datapipes, req_queues, res_queues, call_on_process_init=None):
+    r"""
+    Set the appropriate pipes and protocol server type, and create a loop over multiple datapipes
+    with the protocol server in a non-blocking manner.
+    """
+    assert call_on_process_init is None, "``MultipleDataPipesToQueuesLoop`` does not support call_on_process_init"
+    assert len(source_datapipes) == len(req_queues) and len(req_queues) == len(
+        res_queues
+    ), "``MultipleDataPipesToQueuesLoop`` requires the same number of datapipes, request queues and response queues"
+
+    torch.set_num_threads(1)
+
+    loops = []
+
+    for source_datapipe, req_queue, res_queue in zip(source_datapipes, req_queues, res_queues):
+        # TODO(ejguan): non-blocking request/response introduces non-determinism
+        # because the order of operations is not going to be consistent in this process
```

Comment:
It's actually deterministic because `round_robin_demux` would make sure the order of fetching is preserved.